lib/datadog/tracing/contrib/span_attribute_schema.rb

```diff
@@ -1,26 +1,132 @@
 # frozen_string_literal: true
 
+require_relative 'ext'
+
 module Datadog
   module Tracing
     module Contrib
       # Contains methods for fetching values according to span attributes schema
       module SpanAttributeSchema
         module_function
 
+        def default_span_attribute_schema?
+          Datadog.configuration.tracing.span_attribute_schema ==
+            Tracing::Configuration::Ext::SpanAttributeSchema::DEFAULT_VERSION
+        end
+
+        def active_version
+          case Datadog.configuration.tracing.span_attribute_schema
+          when 'v1'
+            V1
+          else
+            V0 # Default Version
+          end
+        end
+
         def fetch_service_name(env, default)
-          ENV.fetch(env) do
-            if Datadog.configuration.tracing.span_attribute_schema ==
-                Tracing::Configuration::Ext::SpanAttributeSchema::VERSION_ONE
-              Datadog.configuration.service
-            else
+          active_version.fetch_service_name(env, default)
+        end
+
+        def set_peer_service!(span, sources)
+          active_version.set_peer_service!(span, sources)
+        end
+
+        private_class_method :active_version
+
+        # Contains interface of methods to be implemented
+        module Base
+          REFLEXIVE_SOURCES = [Tracing::Metadata::Ext::TAG_PEER_SERVICE].freeze
+          NO_SOURCE = [].freeze
+          IMPLEMENT_ERROR = 'SpanAttributeSchema Version must implement fetch_service_name'
+
+          def self.extended(base)
+            base.private_class_method :not_empty_tag?, :set_peer_service_from_source, :filter_peer_service_sources
```

Comment:
I have to push back on these changes because we are already aware of changes coming down the pipeline which would require separate usage of `v0` and `v1`. Removing the concept entirely here would require a refactor in the very near future.

Originally, we did have differences between `set_peer_service!` but we condensed it into one base class. While it seems excessive at the moment, I think it sets up for the future very nicely and future modules (`v2` onwards) so we should keep it.

`set_peer_service!` does change the internal `span` state too.

Reply:
After talking it over with @delner, we think its best to keep the module structure rather than using classes. We do not want to expose the `Base`. Using the class structure, we increase the surface area of the public API, allowing for things like `Base.set_peer_service!`. We want to be cautious against that so we decided to move forward with our module structure

Reply:
I'm not the biggest fan of this approach, so here's me taking a stab at explaining why. (Note I don't consider this blocking -- e.g. feel free not to use any of this)

1. Private VS Public

> We do not want to expose the Base. Using the class structure, we increase the surface area of the public API, allowing for things like Base.set_peer_service!

Note that the current approach still exposes the Base. For instance, it exposes `Base.set_peer_service!` for everyone to call. If we truly want to make a class private, we need to use `private_constant`

```ruby
module A
  B = 1
  private_constant :B
end
=> A
irb(main):005:0> A::B
(irb):5:in `<main>': private constant A::B referenced (NameError)
```


2. Use of `extended`

```ruby
def self.extended(base)
  base.private_class_method :not_empty_tag?, :set_peer_service_from_source, :filter_peer_service_sources
end
```

The use of `extended` to make method private, seems like a smell, since that functionality could easily be achive by the use of `private` keyworkd. 
Also, it adds inconsistency because hides the method only when the `Base` module is included.

```ruby
module A
  module_function

  def self.extended(base)
    base.private_class_method :hello
  end

  def hello
    puts "hello"
  end
end

module B
  extend A
end
```

So `A.hello` would remain accessible while `B.hello` would be private

```
A.hello
hello
=> nil
irb(main):022:0> B.hello
(irb):22:in `<main>': private method `hello' called for B:Module (NoMethodError)
```

3. The use of `module_function` 

I'm not sure what the advantage of `module_function`` is in this case, as we only want to expose the methods at the singleton level `A.hello`` and the `V0`, `V1``, and `SpanAttributeSchema`` are not meant to be included in other classes or modules. The use of module_function basically converts a method defined for the instance available to the class, which I'm not sure is what we want in this case.

Here is an example of what `module_function` does


```ruby
module A
  module_function

  def hello
    puts "hello"
  end
end
```

Is the same as 

```ruby
module A
  def self.hello
    puts "hello"
  end

  def hello
    puts "hello"
  end
end
```

Putting all of my suggestions together, with the goal of hiding the internal stuff, I suggest keeping the code using modules, but removing the use of `extended`` and `module_function` + hide `Base` using `private_constant`.

```ruby
# frozen_string_literal: true

require_relative 'ext'

module Datadog
  module Tracing
    module Contrib
      # Contains methods for fetching values according to span attributes schema
      module SpanAttributeSchema
        # Contains interface of methods to be implemented
        module Base
          REFLEXIVE_SOURCES = [Tracing::Metadata::Ext::TAG_PEER_SERVICE].freeze
          NO_SOURCE = [].freeze
          IMPLEMENT_ERROR = 'SpanAttributeSchema Version must implement fetch_service_name'

          class << self
            def fetch_service_name(_env, _default)
              raise NotImplementedError, IMPLEMENT_ERROR
            end

            def set_peer_service!(span, sources)
              # Acquire all peer.service values as well as any potential remapping
              peer_service_val, peer_service_source = set_peer_service_from_source(span, sources)
              remap_val = Datadog.configuration.tracing.peer_service_mapping[peer_service_val]

              # Only continue to setting peer.service if actual source is found
              return false unless peer_service_source

              span.set_tag(Tracing::Contrib::Ext::Metadata::TAG_PEER_SERVICE_SOURCE, peer_service_source)

              # Set peer.service to remapped value if found otherwise normally set peer.service
              if remap_val
                span.set_tag(Tracing::Metadata::Ext::TAG_PEER_SERVICE, remap_val)
                span.set_tag(Tracing::Contrib::Ext::Metadata::TAG_PEER_SERVICE_REMAP, peer_service_val)
              else
                span.set_tag(Tracing::Metadata::Ext::TAG_PEER_SERVICE, peer_service_val)
              end
              true
            end

            private

            # set_peer_service_from_source: Implements the extraction logic to determine the peer.service value
            # based on the list of source tags passed as a parameter.
            #
            # If no values are found, it checks the default list for all spans before returning false for no result
            # Sets the source of where the information for peer.service was extracted from
            # Returns a peer.service value if successfully set or not
            def set_peer_service_from_source(span, sources = [])
              # Filter out sources based on existence of peer.service tag
              sources = filter_peer_service_sources(span, sources)

              # Find a possible peer.service source from the list of source tags passed in
              sources.each do |source|
                source_val = span.get_tag(source)
                next unless not_empty_tag?(source_val)

                return source_val, source
              end
              false
            end

            # filter_peer_service_sources: returns filtered sources based on existence of peer.service tag
            # If peer.service exists, we do not read from any other source rather use peer.service as source
            # This is to prevent overwriting of pre-existing peer.service tags
            def filter_peer_service_sources(span, sources)
              ps = span.get_tag(Tracing::Metadata::Ext::TAG_PEER_SERVICE)
              # Do not override existing peer.service tag if it exists based on schema version
              return REFLEXIVE_SOURCES if not_empty_tag?(ps)

              # Check that peer.service is only set on spankind client/producer spans
              if (span.get_tag(Tracing::Metadata::Ext::TAG_KIND) == Tracing::Metadata::Ext::SpanKind::TAG_CLIENT) ||
                  (span.get_tag(Tracing::Metadata::Ext::TAG_KIND) == Tracing::Metadata::Ext::SpanKind::TAG_PRODUCER)
                return sources
              end

              NO_SOURCE
            end

            def not_empty_tag?(tag)
              tag && (tag != '')
            end
          end
        end

        private_constant :Base

        # Contains implementation of methods specific to v0
        module V0
          extend Base

          class << self
            def fetch_service_name(env, default)
              ENV.fetch(env) do
                return Datadog.configuration.service if Datadog.configuration.tracing.global_default_service_name.enabled

                default
              end
            end
          end
        end

        # Contains implementation of methods specific to v1
        module V1
          extend Base

          class << self
            def fetch_service_name(env, _)
              ENV.fetch(env) { Datadog.configuration.service }
            end
          end
        end

        class << self
          def default_span_attribute_schema?
            Datadog.configuration.tracing.span_attribute_schema ==
              Tracing::Configuration::Ext::SpanAttributeSchema::DEFAULT_VERSION
          end

          def fetch_service_name(env, default)
            active_version.fetch_service_name(env, default)
          end

          def set_peer_service!(span, sources)
            active_version.set_peer_service!(span, sources)
          end

          private

          def active_version
            case Datadog.configuration.tracing.span_attribute_schema
            when 'v1'
              V1
            else
              V0 # Default Version
            end
          end
        end
      end
    end
  end
end
```

What do you think @zarirhamza @delner ?

Reply:
A couple of points of clarification:

- > "For instance, it exposes `Base.set_peer_service!` for everyone to call."
   
   I don't believe this is strictly true. `set_peer_service!` is not a `module_function`, it is an instance method of the module. Thus you cannot call this statically. You would have to compose it into a module to access it; I'm not concerned with someone extending it, that's advanced usage.
- > So A.hello would remain accessible while B.hello would be private

    This code is not equivalent: the methods in `Base` are not `module_function` (like in the example), they are instance methods.

Otherwise, your code suggestion (at the bottom of your comment) looks reasonable. The only two things I'd call out as undesirable are (1) its unnecessary for `Base` to expose static-accessible methods, and (2) the use of `class << self` in `V0` and `V1`, which I think is a code smell; particularly it's a `Module` and not a `Class`, and it's... just ugly haha. Functionally I don't have an issue with it.